423 external links media variant

eduardseifert opened the pull request:
Fix #423

# Test URLs

## Before

- https://main--hlx-test--urfuwo.hlx.page/design/design-system
- https://main--hlx-test--urfuwo.hlx.page/design/about

## After

- https://423-external-links-media-variant--hlx-test--urfuwo.hlx.page/design/design-system
- https://423-external-links-media-variant--hlx-test--urfuwo.hlx.page/design/about

# Block

- Block Name: `link-list`
- Documentation: https://423-external-links-media-variant--hlx-test--urfuwo.hlx.page/tools/sidekick/blocks/link-list
- Library Link: https://423-external-links-media-variant--hlx-test--urfuwo.hlx.page/tools/sidekick/library.html?plugin=blocks&path=/tools/sidekick/blocks/link-list&index=0

# Description

- As discussed with @mhaack, this first version replaces the given img element with the corresponding svg markup, so the CSS is capable of manipulating the fill color for hover effects).
- The JS fetches all the icons in the block, replaces the given img tag with the svg markup and strips all the fill attributes, so the CSS can change the color.
- The image is optional.
- 2024-04-15: As discussed: moving on with `img` and CSS filter solution (take a look on the CSS).

saurabh-khare commented:
@eduardseifert Congrats on the first PR :) 
Please provide before and after state for all test Urls so reviewers can check what has changed in the PR. E.g. https://github.com/urfuwo/hlx-test/pull/450 
Also for this specific PR please add link to the page which has link with icon in this block.

eduardseifert commented:
As discuessed with @mhaack, I will move the icons inside the link, change a few lines of CSS and revert the `img` handling to be directly embedded SVG's in the markup (DOM) (insted of referencing them in the `src` attribute).

eduardseifert commented:
- Updated the PR and description
- Removed img to svg transformation
- Using CSS filter for color change
- clean up

mhaack commented on aemedge/blocks/link-list/link-list.js:
```diff
@@ -0,0 +1,30 @@
+ * @param link {Element} The link element
+ */
+function moveLinkTextNodeToSpan(link) {
+  Array.from(link.childNodes)
```
Wait, wrapping in span can be done via decroateIcons, we would not need this function?

eduardseifert replied:
@mhaack `decorateIcons` is used automatically because of the `:icon:` inside the Word content. I need this function, because the icon can be placed on the left and the right side of the link text. Default behavior is plain text + icon (in span). You have no possibility to place margin to the left with text nodes. So I want to span the text to be able to margin to the right.

saurabh-khare commented on aemedge/blocks/link-list/link-list.js:
```diff
@@ -0,0 +1,30 @@
+  Array.from(link.childNodes)
+    .forEach((node) => {
+      if (node.nodeType === Node.TEXT_NODE && node.textContent.trim().length > 0) {
+        const span = document.createElement('span');
```
use dom-builder.js. Also use decorateIcons for wrapping

eduardseifert replied:
Used the dom-builder for span. Thanks. The wrapper itself is needed because the icon can also be placed on the left and right side of the text (need to control the margin).

saurabh-khare commented:
<img width="426" alt="Screenshot 2024-04-15 at 15 50 53" src="https://github.com/urfuwo/hlx-test/assets/41549648/8a4e2c00-787f-4551-8a33-cf40f020a88c">

Infact we don't need the js at all. I just commented it and the default decorateIcons kicks in because you added icons in the content
<img width="262" alt="Screenshot 2024-04-15 at 15 52 16" src="https://github.com/urfuwo/hlx-test/assets/41549648/e3de2163-8287-48e8-843f-56de28b319ac">

saurabh-khare commented:
I compared the existing block with the new block and there are many design changes. Please keep the external links block until we get the feature parity in the new block. Make sure existing content is not impacted.
Reference page: http://localhost:3000/draft/skhare/content-list-test
<img width="1066" alt="Screenshot 2024-04-15 at 16 08 01" src="https://github.com/urfuwo/hlx-test/assets/41549648/52980ff1-c2ae-435e-b2eb-b9a5191eee27">

eduardseifert commented:
@mhaack + @saurabh-khare Please have again a look at the change.

1. Changed to `dom-builder.js` (thanks for the hint).
2. I need the additional `span` for the text to be able to set margin to the left or right (depends on icon position).
3. I have renamed the block as discussed in Slack to `link-list` and changed all the necessary files as well. If you want to have it back to `links-external`, I need to revert it again.
